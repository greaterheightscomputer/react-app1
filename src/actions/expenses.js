import uuid from 'uuid';
import database from '../firebase/firebase';

//Application life cycle in synchrous redux action          //Application life cycle in asynchrous redux action
// component calls action generator                         component calls action generator
// action generator returns object                          action generator returns function
// component dispatches object                              component dispatches function (?)
// redux store changes by running the reducer               function runs (has the ability to dispatch other actions and do whatever it wants)

//4. CREATING THE FOLLOWING ACTION GENETOR FUNCTIONS
// ADD_EXPENSE  
//Synchronous Action Generator Function   
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//Asynchronous Action Generator Function
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description='', 
            note='', 
            amount=0, 
            createdAt=0
        } = expenseData; //destructuring from expenseData
        const expense = { description, note, amount, createdAt };
        
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE    
export const removeExpense = ( { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE  
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// //SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// // asynchronous redux action function
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            // console.log(snapshot.val());
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            }); 

            dispatch(setExpenses(expenses));
        });
    };
};