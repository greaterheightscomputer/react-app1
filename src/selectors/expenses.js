import moment from 'moment';

// 6. Get visible expenses -> this function will match the expenses object with the filters object
// It take in array expenses and destructured filter object and reture filtered and sorted array

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {  
        // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) =>{    //sort() get called on array
          if(sortBy === 'date') {
              return a.createdAt < b.createdAt ? 1 : -1;
        }
        if(sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
      }
    });
};