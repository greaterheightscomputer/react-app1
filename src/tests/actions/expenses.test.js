import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { 
    startAddExpense, 
    addExpense, 
    editExpense, 
    startEditExpense,
    removeExpense, 
    setExpenses, 
    startSetExpenses,
    startRemoveExpense
 } from '../../actions/expenses';
import expenses from '../fixtures/expenses';
import database from '../../firebase/firebase';

const uid = 'thisismytestuid';
const defaultAuthState = { auth: { uid } };
const createMockStore = configureMockStore([thunk]);

beforeEach((done) => {
    const expenseData = {};
    expenses.forEach(({ id, description, note, amount, createdAt }) => {
        expenseData[id] = { description, note, amount, createdAt };
    });
    database.ref(`users/${uid}/expenses`).set(expenseData).then(() => done());
});

test('should setup remove expense action object', () =>{
    const action = removeExpense({ id: '123abc'});
    expect(action).toEqual({
        type: 'REMOVE_EXPENSE',
        id: '123abc'
    })
});

test('should remove expense from firebase', (done) => {
    const store = createMockStore(defaultAuthState);
    const id = expenses[2].id;
    store.dispatch(startRemoveExpense({ id })).then(() => {
        const actions = store.getActions();
        expect(actions[0]).toEqual({
          type: 'REMOVE_EXPENSE',
          id      
        });
        return database.ref(`users/${uid}/expenses/${id}`).once('value');        
        // done();
    })
    .then((snapshot) => {
        expect(snapshot.val()).toBeFalsy();  //when u use snapshot.val() on already remove item it will return null that is why we use toBeFalsy() in the expect() expression.
        done();
    });    
});

test('should setup edit expense action object', () => {
    const action = editExpense('123abc', { note: 'New note value' });
    expect(action).toEqual({
        type: 'EDIT_EXPENSE',
        id: '123abc',
        updates: {
            note: 'New note value'
        }
    });
});

test('should edit expense from firebase', (done) => {
    const store = createMockStore(defaultAuthState);
    const id = expenses[0].id;
    const updates = { amount: 2390 };
    store.dispatch(startEditExpense(id, updates)).then(() => {
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: 'EDIT_EXPENSE',
            id,
            updates
        });
        return database.ref(`users/${uid}/expenses/${id}`).once('value');        
    }).then((snapshot) => {
        expect(snapshot.val().amount).toBe(2390);
        done();
    });
});

test('should setup add expense action object with provided values', () => {    
    const action = addExpense(expenses[2]);
    expect(action).toEqual({
        type: 'ADD_EXPENSE',
        expense: expenses[2]
    });
});

//asynchronous test cases
test('should add expense to database and store', (done) => {
// test('should add expense to database and store', () => { //done() will force jest or test cases to wait untill data is inserted or sink into firebase 
    const store = createMockStore(defaultAuthState); //this createMockStore({}) is a mock or fake 
    const expenseData = {
        description: 'Books',
        amount: 1200,
        note: 'This one is better',
        createdAt: 2000
    };

    //inserting data assertion
    store.dispatch(startAddExpense(expenseData)).then(() => { //then() the then success case will fire after the data as been inserted to the firebase beco's of the return keyword will we add to database.ref('expenses') in the actions/expenses.js after the we will use the return value to make assertion 
        //how we can get data to the redux mock or fake store which will be the return value or data that come back from firebase
        const actions = store.getActions(); //store.getActions() contain all the Action Generator Functions where actions[0] represent addExpense()
        expect(actions[0]).toEqual({
            type: 'ADD_EXPENSE',
            expense: {
                id: expect.any(String),
                ...expenseData            
            }
        });
        
        //fetching data assertion  
        // database.ref(`expenses/${actions[0].expense.id}`).once('value').then((snapshot) =>{ //fetching data from firebase with the id generated from firebase actions[0].expense.id 
        // expect(snapshot.val()).toEqual(expenseData);   
        //     done();     
        // });
        return database.ref(`users/${uid}/expenses/${actions[0].expense.id}`).once('value');  //fetching data from firebase with the id generated from firebase actions[0].expense.id 
    }).then((snapshot) =>{ 
        expect(snapshot.val()).toEqual(expenseData);   
        done();     
    });
});

test('should add expense to with defaults to database and store', (done) => {
// test('should add expense to with defaults to database and store', () => {
    const store = createMockStore(defaultAuthState);
    const expenseData = {
        description: '',
        amount: 0,
        note: '',
        createdAt: 0
    };

    //inserting data assertion
    store.dispatch(startAddExpense({})).then(() => {         
        const actions = store.getActions(); 
        expect(actions[0]).toEqual({
            type: 'ADD_EXPENSE',
            expense: {
                id: expect.any(String),
                ...expenseData            
            }
        });
        
        //fetching data assertion          
        return database.ref(`users/${uid}/expenses/${actions[0].expense.id}`).once('value');  
    }).then((snapshot) =>{ 
        expect(snapshot.val()).toEqual(expenseData);   
        done();     
    });
});

test('should setup set expense action object with data', () => {
    const action = setExpenses(expenses);
    expect(action).toEqual({
        type: 'SET_EXPENSES',
        expenses
    });
    expect(action).toMatchSnapshot();
});

test('should fetch the expenses from firebase', (done) => {
    const store = createMockStore(defaultAuthState);
    store.dispatch(startSetExpenses()).then(() => {
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: 'SET_EXPENSES',
            expenses
        });        
        done();
    });    
});
