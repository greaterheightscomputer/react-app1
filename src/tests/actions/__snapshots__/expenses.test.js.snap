// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should setup set expense action object with data 1`] = `
Object {
  "expenses": Array [
    Object {
      "amount": 195,
      "createdAt": 0,
      "description": "Gum",
      "id": "1",
      "note": "",
    },
    Object {
      "amount": 109500,
      "createdAt": -345600000,
      "description": "Rent",
      "id": "2",
      "note": "",
    },
    Object {
      "amount": 4500,
      "createdAt": 345600000,
      "description": "Credit Card",
      "id": "3",
      "note": "",
    },
  ],
  "type": "SET_EXPENSES",
}
`;
